hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/WindowingProcessors.java

```diff
@@ -47,59 +93,155 @@ private WindowingProcessors() {
-     * the item's {@code eventSeq} to its {@code frameSeq}. Within a frame
-     * items are further classified by a grouping key determined by the {@code
-     * extractKeyF} function. When the processor receives a punctuation with a
-     * given {@code puncSeq}, it emits the current state of all frames with
-     * {@code frameSeq <= puncSeq} and deletes these frames from its storage.
+     * A first stage processor in two-stage sliding window aggregation. See
+     * Two-stage aggregation {@link WindowingProcessors here}.
+     * <p>
+     * This vertex is supposed to be followed by {@link
+     * #slidingWindowStage2(WindowDefinition, WindowOperation)}.
+     *<p>
+     * Output if this vertex is {@link Frame Frame&lt;K, A>}, that is a frame
+     * with a key and accumulator from {@code windowOperation}.
+     *<p>
+     * A frame is identified by its {@code long frameSeq} and {@link
+     * WindowDefinition#higherFrameSeq(long)} maps the item's {@code
+     * eventSeq} to its {@code frameSeq}.
+     * <p>
+     * Within a frame items are further classified by a grouping key determined
+     * by the {@code extractKeyF} function. When the processor receives a
+     * punctuation with a given {@code puncSeq}, it emits the current state of
+     * all frames with {@code frameSeq <= puncSeq} and deletes these frames
+     * from its storage.
      *
      * @param <T> item type
      * @param <K> type of key returned from {@code extractKeyF}
-     * @param <F> type of accumulator returned from {@code windowOperation.
+     * @param <A> type of accumulator returned from {@code windowOperation.
      *            createAccumulatorF()}
      */
-    public static <T, K, F> Distributed.Supplier<Processor> groupByFrame(
-            Distributed.Function<? super T, K> extractKeyF,
-            Distributed.ToLongFunction<? super T> extractEventSeqF,
-            WindowDefinition windowDef,
-            WindowOperation<? super T, F, ?> windowOperation
+    @Nonnull
+    public static <T, K, A> Distributed.Supplier<Processor> slidingWindowStage1(
+            @Nonnull Distributed.Function<? super T, K> extractKeyF,
+            @Nonnull Distributed.ToLongFunction<? super T> extractEventSeqF,
+            @Nonnull WindowDefinition windowDef,
+            @Nonnull WindowOperation<? super T, A, ?> windowOperation
     ) {
-        return () -> new GroupByFrameP<>(extractKeyF, extractEventSeqF, windowDef, windowOperation);
+        // use a single-frame window in this stage; the downstream processor
+        // combines the frames into a window with the user-requested size
+        WindowDefinition tumblingWinDef = new WindowDefinition(
+                windowDef.frameLength(), windowDef.frameOffset(), 1);
+
+        return () -> new WindowingProcessor<T, A, A>(
+                tumblingWinDef,
+                item -> tumblingWinDef.higherFrameSeq(extractEventSeqF.applyAsLong(item)),
+                extractKeyF,
+                WindowOperation.of(
```

Comment:
It's not `a -> a`; it's `(acc, item) -> { biConsumer.accept(acc, item); return biConsumer; }` instead of simply `biConsumer`. You may be right on the point that the separation of the two cases (mutable/immutable) could go even deeper.